go/mysql/binlog_event_compression.go

```diff
@@ -0,0 +1,254 @@
+}
+
+// Create a reader that caches decompressors.
+var zstdDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
+
+type TransactionPayload struct {
+	Size             uint64
+	CompressionType  uint64
+	UncompressedSize uint64
+	Payload          []byte
+	Events           []BinlogEvent
+}
+
+// IsTransactionPayload returns true if a compressed transaction
+// payload event is found (binlog_transaction_compression=ON).
+func (ev binlogEvent) IsTransactionPayload() bool {
+	return ev.Type() == eTransactionPayloadEvent
+}
+
+// TransactionPayload returns the BinlogEvents contained within
+// the compressed transaction.
+// The following event types are compressed as part of the
+// transaction payload:
+//
+//	QUERY_EVENT = 2
+//	INTVAR_EVENT = 5
+//	APPEND_BLOCK_EVENT = 9
+//	DELETE_FILE_EVENT = 11
+//	RAND_EVENT = 13
+//	USER_VAR_EVENT = 14
+//	XID_EVENT = 16
+//	BEGIN_LOAD_QUERY_EVENT = 17
+//	EXECUTE_LOAD_QUERY_EVENT = 18
+//	TABLE_MAP_EVENT = 19
+//	WRITE_ROWS_EVENT_V1 = 23
+//	UPDATE_ROWS_EVENT_V1 = 24
+//	DELETE_ROWS_EVENT_V1 = 25
+//	IGNORABLE_LOG_EVENT = 28
+//	ROWS_QUERY_LOG_EVENT = 29
+//	WRITE_ROWS_EVENT = 30
+//	UPDATE_ROWS_EVENT = 31
+//	DELETE_ROWS_EVENT = 32
+//	XA_PREPARE_LOG_EVENT = 38
+//	PARTIAL_UPDATE_ROWS_EVENT = 39
+//
+// When transaction compression is enabled, the GTID log event has
+// the following fields:
+// +-----------------------------------------+
+// | field_type (1-9 bytes)                  |
+// +-----------------------------------------+
+// | field_size (1-9 bytes)                  |
+// +-----------------------------------------+
+// | m_payload (1 to N bytes)                |
+// +-----------------------------------------+
+// | field_type (1-9 bytes)                  |
+// +-----------------------------------------+
+// | field_size (1-9 bytes)                  |
+// +-----------------------------------------+
+// | m_compression_type (1 to 9 bytes)       |
+// +-----------------------------------------+
+// | field_type (1-9 bytes)                  |
+// +-----------------------------------------+
+// | field_size (1-9 bytes)                  |
+// +-----------------------------------------+
+// | m_uncompressed_size size (0 to 9 bytes) |
+// +-----------------------------------------+
+//
+// We need to extract the compressed transaction payload from the GTID
+// event, decompress it with zstd, and then process the internal events
+// (e.g. Query and Row events) that make up the transaction.
+func (ev binlogEvent) TransactionPayload(format BinlogFormat) ([]BinlogEvent, error) {
+	tp := &TransactionPayload{}
+	if err := tp.Decode(ev.Bytes()[format.HeaderLength:]); err != nil {
+		return nil, vterrors.Wrapf(err, "error decoding transaction payload event")
+	}
+	return tp.Events, nil
+}
+
+// Decode decodes and decompresses the payload.
+func (tp *TransactionPayload) Decode(data []byte) error {
+	log.V(5).Infof("Compressed payload event; length: %d, bytes: %v", len(data), data)
```

Comment:
I think we should remove this log line, as it logs the raw data which could be a PII concern depending on deployments etc. I think we avoid logging raw data across other places as well also for debug level logging?

Reply:
Good point. I'll remove them.